internal/provider/resource_model.go

```diff
@@ -7,326 +7,522 @@ import (
 		if err != nil {
-			return diag.FromErr(err)
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to parse constraints for model, got error: %s", err))
+			return
 		}
-		newConstraints = &aux
 	}
 
-	if d.HasChange("credential") {
-		anyChange = true
-		_, cred := d.GetChange("credential")
-		newCredential = cred.(string)
+	// Check the credential
+	credentialUpdate := ""
+	if !plan.Credential.Equal(state.Credential) {
+		noChange = false
+		credentialUpdate = plan.Credential.ValueString()
 	}
 
-	if !anyChange {
-		return diags
+	if noChange {
+		return
 	}
 
-	cloud := d.Get("cloud").([]interface{})
+	var clouds []nestedCloud
+	resp.Diagnostics.Append(plan.Cloud.ElementsAs(ctx, &clouds, false)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var cloudNameInput string
+
+	if len(clouds) > 0 {
+		cloudNameInput = clouds[0].Name.ValueString()
+	}
 
-	err = client.Models.UpdateModel(juju.UpdateModelInput{
-		UUID:        d.Id(),
-		CloudList:   cloud,
-		Config:      newConfigMap,
+	err = r.client.Models.UpdateModel(juju.UpdateModelInput{
+		UUID:        plan.ID.ValueString(),
+		CloudName:   cloudNameInput,
+		Config:      configMap,
 		Unset:       unsetConfigKeys,
-		Constraints: newConstraints,
-		Credential:  newCredential,
+		Constraints: &newConstraints,
+		Credential:  credentialUpdate,
 	})
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to update model, got error: %s", err))
+		return
 	}
 
-	return diags
+	r.trace(fmt.Sprintf("Updated model resource: %q", plan.Name.ValueString()))
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
 }
 
-// Juju refers to model deletion as "destroy" so we call the Destroy function of our client here rather than delete
-// This function remains named Delete for parity across the provider and to stick within terraform naming conventions
-func resourceModelDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+func (r *modelResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	// Prevent panic if the provider has not been configured.
+	if r.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "model", "delete")
+		return
+	}
 
-	var diags diag.Diagnostics
+	var state modelResourceModel
 
-	modelUUID := d.Id()
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-	err := client.Models.DestroyModel(juju.DestroyModelInput{
-		UUID: modelUUID,
+	err := r.client.Models.DestroyModel(juju.DestroyModelInput{
+		UUID: state.ID.ValueString(),
 	})
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to delete model, got error: %s", err))
+		return
 	}
+	r.trace(fmt.Sprintf("model deleted : %q", state.ID.ValueString()))
```

Comment:
Can we use model name for all of the trace messages please, rather than the UUID.